lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+      newMetricChannel.publish(this);
+    }
+
+    ObjectFreeze(this);
+  }
+
+  /**
+   * The channel name of the metric.
+   * @property {string} channelName
+   */
+  get channelName() {
+    return `metrics:${this.type}:${this.name}`;
+  }
+
+  /**
+   * The channel for this metric.
+   * @property {Channel} channel
+   */
+  get channel() {
+    return this.#channel;
+  }
+
+  /**
+   * Whether the metric should report values. If there are no subscribers,
+   * metric preparation and report construction can be skipped.
+   * @property {boolean} shouldReport
+   */
+  get shouldReport() {
+    return this.#channel.hasSubscribers;
+  }
+
+  /**
+   * Report a value to the metric.
+   * @param {number} value The value to report.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  report(value, meta) {
+    // Skip report construction if there are no subscribers.
+    if (!this.shouldReport) return;
+    const report = new MetricReport(this.type, this.name, value,
+                                    mixMeta(this.meta, meta));
+    this.#channel.publish(report);
+  }
+}
+
+/**
+ * Represents a snapshot of a value in time. Will report the value every time
+ * reset() is called, or when applyDelta() is called with a non-zero value.
+ */
+class Gauge {
+  /**
+   * The metric to report to.
+   * @property {Metric} metric
+   */
+
+  /**
+   * The value of the gauge.
+   * @property {number} value
+   */
+
+  /**
+   * @param {Metric} metric The metric to report to.
+   */
+  constructor(metric) {
+    if (!(metric instanceof Metric)) {
+      throw new ERR_INVALID_ARG_TYPE('metric', ['Metric'], metric);
+    }
+    this.metric = metric;
+    this.value = 0;
+  }
+
+  /**
+   * Set the gauge value.
+   * @param {number} value The value to set the gauge to.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  reset(value = 0, meta) {
+    this.value = value;
+    this.metric.report(value, meta);
+  }
+
+  /**
+   * Apply a delta to the gauge.
+   * @param {number} value The delta to apply to the gauge.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  applyDelta(value, meta) {
+    this.reset(this.value + value, meta);
+  }
+}
+
+/**
+ * Number of events per interval. This will report at every mark() call,
+ * but will report an aggregate value if a sliding window of marks which
+ * occurred within the interval period.
+ */
+class Meter extends Gauge {
+  #window;
+
+  /**
+   * The metric to report to.
+   * @property {Metric} metric
+   */
+
+  /**
+   * The interval in milliseconds to aggregate marks over.
+   * @property {number} interval
+   */
+
+  /**
+   * Construct a new meter.
+   * @param {Metric} metric The metric to report to.
+   * @param {number} interval The interval in milliseconds to aggregate marks over.
+   */
+  constructor(metric, interval) {
+    super(metric);
+    if (typeof interval !== 'number' || interval <= 0) {
```

Comment:
the `interval <= 0` case should probably throw `ERR_INVALID_ARG_VALUE`, likely as a `RangeError`

Reply:
Also, this should probably use the regular validator methods?